src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -722,6 +726,21 @@ private static bool IsTypeLessVisibleThan(TypeSymbol type, Symbol sym, ref Compo
                     }
                 }
 
+                if (visitCustomModifiers && typeWithAnnotationsOpt.HasType)
+                {
+                    foreach (var customModifier in typeWithAnnotationsOpt.CustomModifiers)
+                    {
+                        var result = VisitType(
+                            typeWithAnnotationsOpt: default, type: (NamedTypeSymbol)customModifier.Modifier,
```

Comment:
>(NamedTypeSymbol) [](http://example.com/codeflow?start=67&length=17)

Why the cast is necessary? Is it casting from ```INamedTypeSymbol``` to ```NamedTypeSymbol```? A cast like that is not legal (should never be attempted). I am not sure why it works. Is this code path tested?
BTW, there are only two places where ```CustomModifier.Modifier``` is used in the compiler and only usage in ```SymbolDisplay``` is justified. The one in ```PEFieldSymbol``` though correct, is inefficient. We should be using ```CSharpCustomModifier.ModifierSymbol``` instead.

Reply:
> Is it casting from INamedTypeSymbol to NamedTypeSymbol?

Yes.

> Is this code path tested?

Nope. I recon custom modifiers come from meta data and not from source, what is the correct way to test them? Actually you asked me several times to add code to visit them in the original PR. But the API to get them now is changed.

> BTW, there are only two places where CustomModifier.Modifier is used in the compiler and only usage in SymbolDisplay is justified. The one in PEFieldSymbol though correct, is inefficient. We should be using CSharpCustomModifier.ModifierSymbol instead.

Thanks for the pointer, I'll look into them.

Reply:
BTW, do we have existing tests for `VisitType`?

Reply:
> do we have existing tests for VisitType

A quick search didn't find any direct tests. It is tested indirectly through features that use it. Feel free to add a direct test if you find this way of testing easier.

Reply:
Here is a test scenario that uses a method symbol referring to `T` in custom modifier within containing type. 

The code cannot be executed due to [https://github.com/dotnet/roslyn/issues/58520](https://github.com/dotnet/roslyn/issues/58520). However the containing type of the ```method``` symbol could be used to test ```VisitType``` by calling it directly. Once the bug is fixed, we can add verifications specific to the delegate caching. 
```
        [Fact]
        public void Test1234567()
        {
            var ilSource = @"
.class public auto ansi beforefieldinit C1`1<T>
    extends System.Object
{
    // Methods
    .method public hidebysig static 
        string Method () cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 6 (0x6)
        .maxstack 8

        IL_0000: ldstr ""Method""
        IL_0005: ret
    } // end of method C1`1::Method

    .method public hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x2057
        // Code size 7 (0x7)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void System.Object::.ctor()
        IL_0006: ret
    } // end of method C1`1::.ctor

} // end of class C1`1

.class public auto ansi beforefieldinit C2`1<T>
    extends System.Object
{
    // Methods
    .method public hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x2057
        // Code size 7 (0x7)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void System.Object::.ctor()
        IL_0006: ret
    } // end of method C2`1::.ctor

} // end of class C2`1

.class public auto ansi beforefieldinit C3`1<T>
    extends class C1`1<int32 modopt(class C2`1<!T>)>
{
    // Methods
    .method public hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x205f
        // Code size 7 (0x7)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void class C1`1<int32>::.ctor()
        IL_0006: ret
    } // end of method C3`1::.ctor

} // end of class C3`1
";

            var source = @"
class Test
{
    static void Main()
    {
        M<int>();
    }

    static void M<T>()
    {
        System.Func<string> x = C3<T>.Method;
        System.Console.WriteLine(x());
    }
}
";
            var compilation = CreateCompilationWithIL(source, ilSource, options: TestOptions.ReleaseExe);

            var tree = compilation.SyntaxTrees.Single();
            var model = compilation.GetSemanticModel(tree);
            var method = model.GetSymbolInfo(tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Where(id => id.Identifier.ValueText == "Method").Single()).Symbol.GetSymbol<MethodSymbol>();

            AssertEx.Equal("System.String C1<System.Int32 modopt(C2<T>)>.Method()", method.ToTestDisplayString());

            //CompileAndVerify(compilation, expectedOutput: @"Method");
        }

```

Reply:
Great example! I added this to the PR.

Reply:
> We should be using CSharpCustomModifier.ModifierSymbol instead.

Updated. Please check 67429d3.